perception/lidar_transfusion/lib/preprocess/preprocess_kernel.cu

```diff
@@ -0,0 +1,258 @@
+    int outIndex = current_pillarId * points_per_voxel + i;
+    voxel_features[outIndex * 5] = voxels[inIndex * 5];
+    voxel_features[outIndex * 5 + 1] = voxels[inIndex * 5 + 1];
+    voxel_features[outIndex * 5 + 2] = voxels[inIndex * 5 + 2];
+    voxel_features[outIndex * 5 + 3] = voxels[inIndex * 5 + 3];
+    voxel_features[outIndex * 5 + 4] = voxels[inIndex * 5 + 4];
+  }
+
+  // clear buffer for next infer
+  atomicExch(mask + voxel_index, 0);
+}
+
+// create 4 channels
+cudaError_t PreprocessCuda::generateBaseFeatures_launch(
+  unsigned int * mask, float * voxels, unsigned int * pillar_num, float * voxel_features,
+  unsigned int * voxel_num, unsigned int * voxel_idxs)
+{
+  dim3 threads = {32, 32};
+  dim3 blocks = {divup(config_.grid_x_size_, threads.x), divup(config_.grid_y_size_, threads.y)};
+
+  generateBaseFeatures_kernel<<<blocks, threads, 0, stream_>>>(
+    mask, voxels, config_.grid_y_size_, config_.grid_x_size_, config_.points_per_voxel_,
+    config_.max_voxels_, pillar_num, voxel_features, voxel_num, voxel_idxs);
+  cudaError_t err = cudaGetLastError();
+  return err;
+}
+
+__device__ int decode_field(
+  unsigned char * cloud_data, unsigned long point_offset, unsigned int field_offset,
+  unsigned char field_size, bool is_bigendian)
+{
+  int value = 0;
+  for (int i = 0; i < field_size; i++) {
+    int step = is_bigendian ? field_size - i - 1 : i;
+    value |= cloud_data[point_offset + field_offset + i] << (step * 8);
+  }
+  return value;
+}
+
+__device__ float cast_field(int & field_raw, unsigned char & field_type)
+{
+  switch (field_type) {
+    case 1:  // INT8
+      return static_cast<float>(reinterpret_cast<char &>(field_raw));
+    case 2:  // UINT8
+      return static_cast<float>(reinterpret_cast<unsigned char &>(field_raw));
+    case 3:  // INT16
+      return static_cast<float>(reinterpret_cast<short &>(field_raw));
+    case 4:  // UINT16
+      return static_cast<float>(reinterpret_cast<unsigned short &>(field_raw));
+    case 5:  // INT32
+      return static_cast<float>(reinterpret_cast<int &>(field_raw));
+    case 6:  // UINT32
+      return static_cast<float>(reinterpret_cast<unsigned int &>(field_raw));
+    case 7:  // FLOAT32
+      return reinterpret_cast<float &>(field_raw);
+    case 8:  // FLOAT64
+      return static_cast<float>(reinterpret_cast<double &>(field_raw));
+    default:
+      return 0.0;
+  }
+}
+
+__global__ void generateVoxelsInput_kernel(
+  unsigned char * cloud_data, unsigned int x_offset, unsigned int y_offset, unsigned int z_offset,
+  unsigned int intensity_offset, unsigned char x_datatype, unsigned char y_datatype,
+  unsigned char z_datatype, unsigned char intensity_datatype, unsigned char x_size,
+  unsigned char y_size, unsigned char z_size, unsigned char intensity_size, unsigned int point_step,
+  bool is_bigendian, unsigned int cloud_capacity, unsigned int points_agg, unsigned int points_size,
+  float time_lag, float * affine_past2current, float * points)
```

Comment:
In this context, the transformation does not have the meaning of `affine_past2current` so it makes more sense to have a generic name

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/3c0187f3d1bc8aacc640ea313016cff8b2e7aaf1

Reply:
Sorry If I was not clear enough. The change from `affine_past2current` to `affine_transform` was only in the kernel, where the nuance of transforming between time stamps was lost. :bow:

Reply:
Sorry for misunderstanding
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/8b59a37a133d1d1bdc7b4ae2c606415b60e0f16f